### PR TITLE
Interval unification

### PR DESCRIPTION
**Pull request by abadams**
There were two different conventions for handling unbounded intervals in Solve.cpp and Bounds.cpp. Bounds used undefined Exprs, but Solve needed to distinguish between positive and negative infinity, so used special symbols instead. This PR is a refactor that pulls out the common code.

I went with the convention in Solve.cpp. Hopefully it makes Bounds.cpp more readable.

### CODE REVIEW

File: src/Bounds.cpp [hunk 1]
```diff
@@ -99,61 +99,59 @@ class Bounds : public IRVisitor {
...
     void visit(const Cast *op) {
 
         op->value.accept(this);
-        Expr min_a = min, max_a = max;
+        Interval a = interval;
 
-        if (min_a.same_as(op->value) && max_a.same_as(op->value)) {
-            min = max = op;
+        if (a.is_single_point(op->value)) {
+            interval = Interval(op, op);
```
Comment (zvookin) [inline src/Bounds.cpp]: What guides the choice to use "Interval(op, op)" instead of "Interval::single_point(op)"?

File: src/Bounds.cpp [hunk 2]
```diff
@@ -166,29 +164,28 @@ class Bounds : public IRVisitor {
...
-                test = simplify(test);
-                if (is_one(test)) {
+                if (can_prove(test)) {
                     could_overflow = false;
                     // Relax the bounds to the constants we found. Not
                     // strictly necessary, but probably helpful to
                     // keep the expressions small.
-                    min_a = lower_bound;
-                    max_a = upper_bound;
+                    a.min = lower_bound;
```
Comment (zvookin) [inline src/Bounds.cpp]: I think stylistically, assigning Interval(lower_bound, upper_bound) is clearer.

File: src/Bounds.cpp [hunk 3]
```diff
@@ -197,11 +194,11 @@ class Bounds : public IRVisitor {
             // Start with the bounds of the narrow type.
             bounds_of_type(from);
             // If we have a better min or max for the arg use that.
-            if (min_a.defined()) min = min_a;
-            if (max_a.defined()) max = max_a;
+            if (a.has_lower_bound()) interval.min = a.min;
```
Comment (zvookin) [inline src/Bounds.cpp]: I feel there may be room for an Interval method to copy the bounds which are defined in one interval to another interval.

File: src/Bounds.cpp [hunk 4]
```diff
@@ -210,464 +207,354 @@ class Bounds : public IRVisitor {
...
-                Expr a = min_a * min_b;
-                Expr b = min_a * max_b;
-                Expr cmp = min_a >= make_zero(min_a.type().element_of());
-                min = select(cmp, a, b);
-                max = select(cmp, b, a);
-            }
-        } else if (min_b.same_as(max_b)) {
-            // B is constant
-            if (is_zero(min_b)) {
-                min = max = min_a;
```
Comment (zvookin) [inline src/Bounds.cpp]: This was a bug right?

Comment (abadams) [inline src/Bounds.cpp]: It was!

File: src/Simplify.cpp [hunk 5]
```diff
@@ -3765,6 +3765,10 @@ Stmt simplify_exprs(Stmt s) {
     return SimplifyExprs().mutate(s);
 }
 
+bool can_prove(Expr e) {
```
Comment (zvookin) [inline src/Simplify.cpp]: I'm a little worried about this being called for non-Bool expressions and some one expecting it to return true if the Expr is non-zero. Perhaps an assert that the type of the passed in Expr is Bool?

**Comment by zvookin** [comment]
LGTM.

File: src/TrimNoOps.cpp [hunk 6]
```diff
@@ -199,13 +199,18 @@ class SimplifyUsingBounds : public IRMutator {
                 break;
             } else if (!expr_uses_var(test, loop.var)) {
                 continue;
-            }  else if (is_one(simplify(loop.i.min == loop.i.max)) && expr_uses_var(test, loop.var)) {
+            }  else if (loop.i.is_bounded() &&
+                        can_prove(loop.i.min == loop.i.max) &&
+                        expr_uses_var(test, loop.var)) {
+                //debug(0) << "Substituting out " << loop.var << "\n";
```
Comment (dsharletg) [inline src/TrimNoOps.cpp]: Commented out debugging code?

File: src/Bounds.cpp [hunk 7]
```diff
@@ -99,61 +99,59 @@ class Bounds : public IRVisitor {
...
         }
     }
 
     void bounds_of_type(Type t) {
         t = t.element_of();
         if (t.is_uint() && t.bits() <= 16) {
-            max = cast(t, (1 << t.bits()) - 1);
-            min = cast(t, 0);
+            interval = Interval(cast(t, 0),
+                                cast(t, (1 << t.bits()) - 1));
```
Comment (dsharletg) [inline src/Bounds.cpp]: t.max()?

File: src/Bounds.cpp [hunk 8]
```diff
@@ -99,61 +99,59 @@ class Bounds : public IRVisitor {
...
+
         } else if (t.is_int() && t.bits() <= 16) {
-            max = cast(t, (1 << (t.bits()-1)) - 1);
-            min = cast(t, -(1 << (t.bits()-1)));
+            interval = Interval(cast(t, -(1 << (t.bits()-1))),
+                                cast(t, (1 << (t.bits()-1)) - 1));
```
Comment (dsharletg) [inline src/Bounds.cpp]: t.min() and t.max()?

**Comment by abadams** [comment]
merging this and several other outstanding things locally, so I can test their combination for build breakages before pushing

**Pull request merged**